osu.Server.Spectator/Hubs/MultiplayerHub.cs

```diff
@@ -584,24 +550,60 @@ private void ensureIsHost(MultiplayerRoom room)
         /// <summary>
         /// Retrieve the <see cref="MultiplayerRoom"/> for the local context user.
         /// </summary>
-        private async Task<MultiplayerRoom> getLocalUserRoom()
+        private async Task<ItemUsage<MultiplayerRoom>> getLocalUserRoom(MultiplayerClientState? state)
         {
-            var state = await GetLocalUserState();
-
             if (state == null)
                 throw new NotJoinedRoomException();
 
             long roomId = state.CurrentRoomID;
 
-            MultiplayerRoom? room;
+            return await ACTIVE_ROOMS.GetForUse(roomId);
+        }
 
-            lock (active_rooms)
+        private async Task leaveRoom(MultiplayerClientState state)
+        {
+            using (var roomUsage = await getLocalUserRoom(state))
             {
-                if (!active_rooms.TryGetValue(roomId, out room))
-                    failWithInvalidState($"User ({CurrentContextUserId}) is in a room ({roomId}) this hub is not aware of.");
-            }
+                var room = roomUsage.Item;
+
+                if (room == null)
+                    throw new InvalidOperationException("Attempted to operate on a null room");
```

Comment:
I think throwing anything in `leaveRoom()` as is the case here can cause the user state to not be destroyed higher up in `LeaveRoom()` while it should probably be. Should most likely catch errors of this method higher up, destroy and re-throw. (in a finally block maybe?)

Also, similarly, `CleanUpState` probably shouldn't care about any exceptions thrown here? Not 100% sure on that one.

Reply:
Moved to finally block to hopefully avoid this.